### DISCUSSION TITLE
`docker run` alternative

### DISCUSSION THREAD

**Discussion by remal** [testcontainers/testcontainers-java#5260] [Q&A] [answered]
I'd like to execute a command like this: `docker run schemaspy/schemaspy <options>`. Is it possible to do it via Testcontainers?

If I understand correctly, Testcontainers requires me to start the container first and then execute a command inside a running container. However, I don't want to start the container - I'd like to execute a command in the container **synchronously** and get a result.

I suppose I have to "disable" `ENTRYPOINT` and then call `execInContainer()`. If it is so, could you please explain how to "disable" `ENTRYPOINT`? Replacing `ENTRYPOINT` with some dummy command, like `touch` is not an option, as a container can be "distroless".

**Comment by bsideup** [comment, marked as answer]
Hi @remal,

Have a look at `OneShotStartupCheckStrategy`:
https://www.testcontainers.org/features/startup_and_waits/#one-shot-startup-strategy-example

**Reply by remal** [reply]
Hi @bsideup,

Thanks for suggesting the strategy!

Is it possible to remove `ENTRYPOINT` at all? I'd like the container to start without any command being executed and then call `execInContainer()`.

**Reply by kiview** [reply]
Yes, this is possible using `withCreateContainerCmdModifier`:
https://www.testcontainers.org/features/advanced_options/#customizing-the-container

See the `docker-java` API:
https://github.com/docker-java/docker-java/blob/3.2.1/docker-java-api/src/main/java/com/github/dockerjava/api/command/CreateContainerCmd.java#L106

**Comment by SimonHarmonicMinor** [comment]
@remal Have you managed to run schemaSpy with Testcontainers? If so, can you share the setup, please? I'm also trying to do it but now I'm stuck. Here is something I came up with:

```java
private final GenericContainer<?> SCHEMA_SPY =
        new GenericContainer<>(DockerImageName.parse("schemaspy/schemaspy:6.1.0"))
            .withNetworkAliases("schemaspy")
            .withCreateContainerCmdModifier(cmd -> cmd.withEntrypoint(""))
            .withCommand("sh", "-c", "sleep 50000 && echo \"hello\"", "HELLO_TESTCONTAINERS")
            .withNetwork(NETWORK)
            .withStartupCheckStrategy(
                new IndefiniteWaitOneShotStartupCheckStrategy()
            )
            .withLogConsumer(new Slf4jLogConsumer(LoggerFactory.getLogger("SchemaSpy")));

    @Test
    @SneakyThrows
    void test() {
        Startables.deepStart(SCHEMA_SPY);
        while (!SCHEMA_SPY.isRunning()) {
            Thread.yield();
        }
        SCHEMA_SPY.execInContainer(
            "java -jar schemaspy-6.1.0.jar -t pgsql11 -db %s -host postgres -u %s -p %s -debug && tail -f /dev/null"
                .formatted(POSTGRES.getDatabaseName(), POSTGRES.getUsername(), POSTGRES.getPassword())
        );
        SCHEMA_SPY.execInContainer("tar", "-czvf", "/output/output.tar.gz", "/output");
        SCHEMA_SPY.copyFileFromContainer(
            "/output/output.tar.gz",
            Path.of(getClass().getResource("/").getPath(), "output.tar.gz")
                .toAbsolutePath()
                .toString()
        );

        SCHEMA_SPY.stop();
    }
```

Here is the idea:

1. I set the schemaSpy container with `sleep 50000` command so the container doesn't quit immediately after start.
1. Then I run it asynchronously with `Startable.deepStart` and wait until it's marked as RUNNING in the while loop.
1. Finally, I run the command to generate the schema with `execInContainer`, archive the result folder structure into tarball, and copy it from the container to the OS.

But actually the `execInContainer` command fails with the message

```
execInContainer can only be used while the Container is running
```

So, the `SCHEMA_SPY.isRunning()` method returns `true`, but the container is not running actually. Any ideas about this problem?

**Reply by remal** [reply]
@SimonHarmonicMinor, yes, I have.

I wrote something like this:
```java
try (var network = Network.newNetwork()) {
    try (var databaseContainer = new PostgreSQLContainer<>()) {
        databaseContainer.setNetwork(network);
        databaseContainer.setNetworkAliases(singletonList("postgresql"));

        databaseContainer.start();

        var dataSource = newDataSource(databaseContainer);
        executeFlywayMigrations(dataSource);


        var properties = new Properties();
        properties.setProperty("schemaspy.t", "pgsql11");
        properties.setProperty("schemaspy.host", databaseContainer.getNetworkAliases().get(0));
        properties.setProperty("schemaspy.port", String.valueOf(5432));
        properties.setProperty("schemaspy.db", databaseContainer.getDatabaseName());
        properties.setProperty("schemaspy.u", databaseContainer.getUsername());
        properties.setProperty("schemaspy.p", databaseContainer.getPassword());
        var propertiesFile = new File(tmpDir, "schemaspy.properties");
        try (var outputStream = newOutputStream(propertiesFile.toPath())) {
            properties.store(outputStream, null);
        }
        
        try (var container = new GenericContainer<>(DockerImageName.parse("schemaspy/schemaspy:latest"))) {
            container.setNetwork(databaseContainer.getNetwork());
            container.addFileSystemBind(
                    propertiesFile.getAbsolutePath(),
                    "/schemaspy.properties",
                    READ_ONLY
            );
            container.addFileSystemBind(
                    getOutputDir().getAbsolutePath(),
                    "/output",
                    READ_WRITE
            );
            if (schemaMetaFile != null) {
                container.addFileSystemBind(
                        schemaMetaFile.getAbsolutePath(),
                        "/schemameta.xml",
                        READ_ONLY
                );
            }
            container.setLogConsumers(singletonList((OutputFrame frame) ->
                    getLogger().info(frame.getUtf8String())
            ));
            container.setWaitStrategy(new LogMessageWaitStrategy().withRegEx(Stream.of(
                    "View the results by opening",
                    "Failed to connect to database URL"
            ).collect(joining("|", "(?i).*(", ").*"))));
            container.start();
        }
    }
}
```

**Reply by SimonHarmonicMinor** [reply]
@remal Thank you for sharing this piece. Unfortunately, it didn't work for me. The container is stuck, if I add the `addFileSystemBind` configuration. Anyway, I found another solution. I posted the entire explanation on [this StackOverflow question](https://stackoverflow.com/a/75651041/9167731), but here are the main points:

1. Start the SchemaSpy container and override the `entrypoint` with `sleep 500000`. So, it doesn't stop automatically.
1. Execute the documentation generation command with `execInContainer` method.
1. Execute the archiving of result folder structure in tarball inside the container.
1. Copy the tarball from the container to the OS directory.

Now you can unpack the archive and host it on GitHub/GitLab pages
